src/stomp_moveit_planning_context.cpp

```diff
@@ -109,10 +109,13 @@ bool StompPlanningContext::solve(planning_interface::MotionPlanResponse& res)
   const auto task = createStompTask(config, *this);
   stomp_ = std::make_shared<stomp::Stomp>(config, task);
 
-  // Timeout async task
+  std::condition_variable cv;
+  std::mutex cv_mutex;
+  bool finished;
   auto timeout_future = std::async(std::launch::async, [&, stomp = stomp_]() {
-    std::this_thread::sleep_for(std::chrono::duration<double>(req.allowed_planning_time));
-    if (stomp)
+    std::unique_lock<std::mutex> lock(cv_mutex);
+    cv.wait_for(lock, std::chrono::duration<double>(req.allowed_planning_time), [&finished]{return finished;});
+    if (stomp_ == stomp)
```

Comment:
If we run into this case, we might have a problem. The first check was really only checking for pointer reset (which means planning ended anyway). Now it might be possible to invalidate previous timeouts by running new planning requests (stomp_ changes). That's an edge case, but we should probably switch to only using a locally scoped STOMP instance, or to synchronize the solve() function.

Reply:
there should be a new async task for every new stomp_ instance. I'm just thinking that you still want to cancel an "old" stomp instance even if you have already started planning with a new one, which technically is still possible at this point. https://github.com/ros-planning/stomp_moveit/pull/9#discussion_r1150481973 should take care of that situation at least.

atomic pointers would also not really help probably, since they could also be invalidated by a new call to `solve()`

Reply:
The only way to really make the API consistent would be to reject new calls to `solve()` as long as there is an active plan